lib/mayaUsd/nodes/proxyShapeBase.h

```diff
@@ -358,6 +373,19 @@ class MayaUsdProxyShapeBase
 
     // Whether or not the proxy shape has enabled UFE/subpath selection
     const bool _isUfeSelectionEnabled;
+
+    // For unshared composition
+    SdfLayerRefPtr _unsharedStageRootLayer;
+
+    // We need to keep track of unshared sublayers (otherwise they get removed)
+    std::vector<SdfLayerRefPtr> _unsharedStageRootSublayers;
+
+    // Keep track of the incoming layers
+    std::set<std::string> _incomingLayers;
+
+    // Keep track of source stage's root identifier (when we change the source we can swap it
+    // in-place in the layer heirarchy)
+    std::string _sourceRootIdentifier;
```

Comment:
This is because I need to keep track of the source. On the proxy shape the source can be one of 4 things:

1. Stage Data In port
2. Stage Cache ID In Port
3. File Path (Valid)
4. New Anon Stage (if all else fails)

So when you are in unshared mode that source above is greyed out but you can still move it in the hierarchy. Meaning you can have

UnsharedRootLayer
|---> Some User Created Middle Layer
   |---> The source (referenced above, buried in a complex hierarchy)

So now when we change the file or change the source we want to maintain that hierarchy, so we need to keep track of what the source was and what it will be. 

After our discussion yesterday I was trying to access the old stage from the out port, but its hard for me to figure out which layer is which, the code gets really complicated if you don't want to keep track of the source layer. Simplest thing is to keep it as is.

Reply:
Done